src/backend/access/zedstore/zedstore_tidpage.c

```diff
@@ -1380,6 +1380,17 @@ zsbt_tid_add_items(Relation rel, Buffer buf, List *newitems, zs_pending_undo_op
 		OffsetNumber startoff;
 		OffsetNumber off;
 
+		if (RelationNeedsWAL(rel))
+		{
+			XLogBeginInsert();
+			/*
+			 * We allocate an rdata per tid leaf item. We need two extra rdatas for UNDO
+			 * This must be called before we enter the critical section as
+			 * XLogEnsureRecordSpace performs memory allocation.
+			 */
+			XLogEnsureRecordSpace(0, list_length(newitems) + 1 + 2);
```

Comment:
I would like if you can spit these changes into separate commit. They are not directly connected with reserve tids, but are required in-general with prior code as well.

Reply:
Makes sense!